app/Models/Achievement.php

```diff
@@ -192,6 +194,22 @@ public function getActivitylogOptions(): LogOptions
 
     // == helpers
 
+    public function upsertAuthorshipCredit(User $user, AchievementAuthorTask $task, ?Carbon $backdate = null): AchievementAuthor
+    {
+        return $this->authorshipCredits()->updateOrCreate(
+            ['user_id' => $user->id, 'task' => $task->value],
+            ['created_at' => $backdate ?? now(), 'updated_at' => now()]
+        );
+    }
+
+    public function removeAuthorshipCredit(User $user, AchievementAuthorTask $task): int
+    {
+        return $this->authorshipCredits()
+            ->where('user_id', $user->id)
+            ->where('task', $task->value)
+            ->delete();
```

Comment:
There's probably a valid reason to allow it, but should we be allowed to remove the credit entry for the user associated to the achievement? i.e. can I remove my logic credit from an achievement that I uploaded?

Reply:
Kind of a forward-thinking question: Should dummy records be created for all tasks using the achievement author and created timestamp? It's implied that the original author is credited with the logic, artwork, writing, and testing until such time as someone else is given credit.

Instead of inserting dummy records for the all of those tasks, we could minimize the size of the table and just assume those values (this would make more sense for a user-facing UI, and not the management UI). An actual value would have to be inserted when another user is credited for historical purposes, but that could be deleted if the user didn't actually participate in that task.

Reply:
I'm not sure why I even added this `removeAuthorshipCredit()` function, but it appears to be unused at the moment. I've deleted it in latest.

You've raised a good point; I think we need elevated privileges around soft deletion. In latest, only `DEVELOPER_STAFF`, `MODERATOR`, and `TEAM_ACCOUNT` can perform soft deletes.

Reply:
> Should dummy records be created for all tasks using the achievement author and created timestamp? It's implied that the original author is credited with the logic, artwork, writing, and testing until such time as someone else is given credit.

Given the current implication, I lean towards no.

> Instead of inserting dummy records for the all of those tasks, we could minimize the size of the table and just assume those values (this would make more sense for a user-facing UI, and not the management UI).

Agreed - this is where I'm currently at. We can run with the assumption unless a value for some task with a different user ID is inserted.

> An actual value would have to be inserted when another user is credited for historical purposes, but that could be deleted if the user didn't actually participate in that task.

I've made an adjustment for this in latest. Now when adding non-logic credits to an achievement, the Filament relation manager will try to backfill data

Reply:
I'm still able to delete credits for achievements I've created.
![image](https://github.com/user-attachments/assets/39f0b3b2-cfcc-4c3a-bd83-3fdc51afba0d)

Reply:
As someone with either a DEVELOPER_STAFF, MODERATOR, or TEAM_ACCOUNT role, this is expected (`AchievementAuthorPolicy::delete()`). Should we further refine these restrictions?

This is what it looks like for a developer without one of those three roles:
![Screenshot 2024-09-22 at 10 59 46 AM](https://github.com/user-attachments/assets/34eb0cb4-3821-4226-bf72-14a369429e3c)

Reply:
It's a valid concern.

In latest, the delete action is hidden for the earliest logic credit associated with the achievement.